modules/module-postgres-storage/src/storage/PostgresSyncRulesStorage.ts

```diff
@@ -0,0 +1,664 @@
+      .first();
+
+    if (syncRulesRow == null) {
+      throw new Error('Cannot find sync rules status');
+    }
+
+    return {
+      snapshot_done: syncRulesRow.snapshot_done,
+      active: syncRulesRow.state == storage.SyncRuleState.ACTIVE,
+      checkpoint_lsn: syncRulesRow.last_checkpoint_lsn ?? null
+    };
+  }
+
+  async clear(): Promise<void> {
+    await this.db.sql`
+      UPDATE sync_rules
+      SET
+        snapshot_done = FALSE,
+        last_checkpoint_lsn = NULL,
+        last_checkpoint = NULL,
+        no_checkpoint_before = NULL
+      WHERE
+        id = ${{ type: 'int4', value: this.group_id }}
+    `.execute();
+
+    await this.db.sql`
+      DELETE FROM bucket_data
+      WHERE
+        group_id = ${{ type: 'int4', value: this.group_id }}
+    `.execute();
+
+    await this.db.sql`
+      DELETE FROM bucket_parameters
+      WHERE
+        group_id = ${{ type: 'int4', value: this.group_id }}
+    `.execute();
+
+    await this.db.sql`
+      DELETE FROM current_data
+      WHERE
+        group_id = ${{ type: 'int4', value: this.group_id }}
+    `.execute();
+
+    await this.db.sql`
+      DELETE FROM source_tables
+      WHERE
+        group_id = ${{ type: 'int4', value: this.group_id }}
+    `.execute();
+  }
+
+  async autoActivate(): Promise<void> {
+    await this.db.transaction(async (db) => {
+      const syncRulesRow = await db.sql`
+        SELECT
+          state
+        FROM
+          sync_rules
+        WHERE
+          id = ${{ type: 'int4', value: this.group_id }}
+      `
+        .decoded(pick(models.SyncRules, ['state']))
+        .first();
+
+      if (syncRulesRow && syncRulesRow.state == storage.SyncRuleState.PROCESSING) {
+        await db.sql`
+          UPDATE sync_rules
+          SET
+            state = ${{ type: 'varchar', value: storage.SyncRuleState.ACTIVE }}
+          WHERE
+            id = ${{ type: 'int4', value: this.group_id }}
+        `.execute();
+      }
+
+      await db.sql`
+        UPDATE sync_rules
+        SET
+          state = ${{ type: 'varchar', value: storage.SyncRuleState.STOP }}
+        WHERE
+          state = ${{ type: 'varchar', value: storage.SyncRuleState.ACTIVE }}
+          AND id != ${{ type: 'int4', value: this.group_id }}
+      `.execute();
+    });
+  }
+
+  private async getChecksumsInternal(batch: storage.FetchPartialBucketChecksum[]): Promise<storage.PartialChecksumMap> {
+    if (batch.length == 0) {
+      return new Map();
+    }
+
+    const rangedBatch = batch.map((b) => ({
+      ...b,
+      start: b.start ?? 0
+    }));
+
+    const results = await this.db.sql`
+      WITH
+        filter_data AS (
+          SELECT
+            FILTER ->> 'bucket' AS bucket_name,
+            (FILTER ->> 'start')::BIGINT AS start_op_id,
+            (FILTER ->> 'end')::BIGINT AS end_op_id
+          FROM
+            jsonb_array_elements(${{ type: 'jsonb', value: rangedBatch }}::jsonb) AS FILTER
+        )

```

Comment:
This approach was mentioned as a performance issue for `getBucketDataBatch`, but not here? Is it just because these queries are not run as often?

Probably fine for the intial release, but we should properly investigate later.

Reply:
I suspect we probably could get some performance boost from changing this to the multiple "OR" method, but I haven't done those levels of optimisation yet.